src/libertem/udf/raw.py

```diff
@@ -75,3 +76,107 @@ def merge_all(self, ordered_results):
         return {'intensity': intensity_sum}
+
+
+class PickShiftedUDF(UDF):
+    '''
+    Load raw data from ROI and applies a corrective shift.
+
+    This UDF is useful to compensate for descan, separating the
+    effect from material characteristics.
+
+    Parameters
+    ----------
+    regression_coefficients : (3, 2) matrix
+        A vector of regression coefficients for x- and y-directions.
+        The regression coefficients make up two planes that describe
+        the descan and are used to remove it.
+        The coefficients are arranged in the following manner:
+        ((y, x), (dy/dy, dx/dy), (dy/dx, dx/dx)) where y and x are
+        the shifts at coordinates (0, 0). The displayed image will be
+        shifted in x- and y-directions by these parameters. By default
+        `None`, which corresponds to no shifts.
+    '''
+    def __init__(self, regression_coefficients=None):
+        if regression_coefficients is None:
+            regression_coefficients = np.zeros((3, 2))
+
+        super().__init__(regression_coefficients=regression_coefficients)
+
+    def process_tile(self, tile):
+        ''
+        # We work in flattened nav space with ROI applied
+        sl = self.meta.slice.get()
+        self.results.intensity[sl] = tile
+        self.results.coordinates[:] = self.meta.coordinates
+
+    def merge(self, dest, src):
+        ''
+        # We receive full-size buffers from each node that
+        # contributes at least one frame and rely on the rest being filled
+        # with zeros correctly.
+        dest.intensity[:] += src.intensity
+        dest.coordinates[:] = src.coordinates
+
+    def merge_all(self, ordered_results):
+        ''
+        intensity_chunks = [b.intensity for b in ordered_results.values()]
+        intensity_sum = np.stack(intensity_chunks, axis=0).sum(axis=0)
+        coordinates = np.concatenate([b.coordinates for b in ordered_results.values()], axis=0)
+        return {'intensity': intensity_sum, "coordinates": coordinates}
+
+    def get_results(self):
+
+        intensity = self.results.get_buffer('intensity').data
+
+        regression_coefficients = np.array(self.params.regression_coefficients)
+        # Only do shifts if nonzero regression coefficients
+        if regression_coefficients.any():
+
+            coordinates = self.results.get_buffer('coordinates').raw_data
+
+            prefix_ones = np.ones((*coordinates.shape[:-1], 1))
+            coordinates = np.concatenate((prefix_ones, coordinates), axis=-1)
+            shifts = np.dot(coordinates, self.params.regression_coefficients)
+
+            shifted = []
+            for intens, shift in zip(intensity, shifts):
+                shifted.append(scipy.ndimage.shift(intens, -shift, mode="constant"))
+
+            if shifted:
+                intensity = np.stack(shifted)
+
+        return {
+            "intensity": intensity,
+            "coordinates": self.results.get_buffer('coordinates').raw_data
+        }
```

Comment:
What about using `process_frame()` and doing all this there? It would be easier to read, and the processing would be performed on the workers, not on the main process.

Later one could look into using `process_partition()` or `postprocess()` to process larger chunks, which would be beneficial for using GPUs.